rfcs/rfc0019.md

```diff
@@ -0,0 +1,174 @@
+# A short and accurate title
+
+## Preamble
+
+    Author:  Ricardo Signes <rjbs@semiotic.systems>
+    Sponsor: RJBS
+    ID:      RFC 0018
+    Status:  Draft
+
+## Abstract
+
+This document proposes a new quote-like operator for string literals.  This
+operator, `qt`, is meant to serve as an alternative to `qq`, with simpler rules
+for understanding what is interpolated and how.  It takes its design from
+JavaScript's template literals and Ruby's string interpolation.
+
+Instead of allowing a subset of variable expressions directly within a string,
+qt allows *any* expression to be interpolated when delimited within the string
+literal.
+
+## Motivation
+
+Existing interpolating strings are extremely convenient, but have shortcomings
+built into their design.  Examples:
+
+```perl
+my $what   = "balloons";
+my $colors = [ qw( red white blue ) ];
+
+my $object   = Party->new({ type => "birthday", bring => $what });
+
+$s = "$problems";     # "" . $problems
+$s = "$colors->[0]";  # "" . $colors->[0]
+
+$s = "@$colors";      # join($", @$colors)
+$s = "$colors->@*";   # If postderef_qq on:  join($", $colors->@*)
+                      # Otherwise         :  "" . $colors . "->@*"
+
+$s = "Bring $object->{bring}";  # "Bring " . $object->{bring}
+$s = "Bring $object->bring";    # "Bring " . $object . "->bring}"
+```
+
+As we look into adding new forms of dereference (for example the proposed `?->`
+operator), or at figuring out how to make `postderef_qq` a default behavior, we
+are likely to keep hitting confusing problems with the behavior of `qq`.  We
+should add something that designs away all these problems now and for the
+future.
+
+## Rationale
+
+The proposed `qt` operator only looks for one special token in a string
+literal: `{`.  Source content until the matching `}` is treated as a scalar
+expression to be interpolated into the string in place of the `{...}`
+construct.  This is arbitrarily extensible for other existing expressions and
+for new ones that may be added.
+
+## Specification
+
+This:
+
+```
+qt{A { TEXT } B};
+```
+
+will be equivalent to
+
+```
+qq{A ${ \scalar TEXT } B};
```

Comment:
I get where you're going with this, but a strict pedant would point out that there might be operator precedence problems here. Strictly speaking it should be equivalent to

    qq{A ${ \scalar ( TEXT ) } B};

because `TEXT` might contain something like `C() or D()`, which would parse lower than the `scalar` operator.

Reply:
would it be safe to think of `qt(STRING)` as an optimized (and legible) shortcut to:

```perl
STRING =~ s:{(.*?)}:${ \scalar($1) }:eeg;
```
?

Reply:
@garu "Kinda, yes" I think.  I am not going to try to think very hard about edge cases.  Roughly that.